### PR TITLE
Import ISLET project for Arm CCA

### PR DESCRIPTION
**Pull request by bitboom**
First, I am glad to collaborate with `Certifier` :)

This PR imports `ISLET SDK` and adds the sample application.
This version runs on x86_64 to integrate easily.

My plan is like:
1. Import ISLET SDK and make the sample app (simulated version on x86_64) << **This PR**
2. Integrate interfaces between `ceritifer` and `islet` (x86_64) << *I hope that certifier developer could help this!*
3. Run sample app on Arm CCA simulator (arm64)

## ISLET SDK
It is for applications running on Arm CCA. ISLET SDK is written in `rust` but we also support `C/C++`.
We use `cbindgen` which is the powerful tool to create headers for rust libraries which expose public C/C++ APIs.
_So please do not modify header directly._

Please refer [this](https://github.com/Samsung/islet/tree/main/sdk) for details.

```
+--------+                           +---------+
| sdk    | => cbindgen => header  => | app     |
| (rust) | => cdylib   => library => | (c/c++) |
+--------+                           +---------+
```

Feel free to give me feedback and any questions.

### CODE REVIEW

**Comment by bitboom** [comment]
I am trying to find proper the open-source manager in my company. After that, I am going to sign the `contributor license agreement`.

File: sample_apps/islet_attest_seal/README.md [hunk 1]
```diff
@@ -0,0 +1,21 @@
+# A sample app towards Arm CCA
+This sample app is towards Arm CCA with [ISLET SDK](https://github.com/Samsung/islet/tree/main/sdk)
+Currently this is running on x86_64 to intergrate easily.
```
Comment (gapisback) [inline sample_apps/islet_attest_seal/README.md:3]: Typo: 'integrate'

File: sample_apps/islet_attest_seal/attest_seal_app.cc [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+}
+
+int main() {
+    printf("[!] Starting attestation test...\n");
+    if (attestation_test()) 
+        printf("[!] Success to attestation test...\n");
+    else
+        printf("[!] Failed to attestation test...\n");
+
+    printf("[!] Starting sealing test...\n");
```
Comment (gapisback) [inline sample_apps/islet_attest_seal/attest_seal_app.cc:72]: Qs here: If `attestation_test()` were to FAIL, why would we want to continue wiht sealing test? 

Do we not want to exit if `attestation_test()`  returns fals?

File: sample_apps/islet_attest_seal/attest_seal_app.cc [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+    if (sealing_test()) 
+        printf("[!] Success to sealing test...\n");
+    else
+        printf("[!] Failed to sealing test...\n");
```
Comment (gapisback) [inline sample_apps/islet_attest_seal/attest_seal_app.cc:76]: Please rewrite this block as suggested previously, using `bool rv` and ternary for `printf()`.

File: third_party/islet/include/islet.h [hunk 4]
```diff
@@ -0,0 +1,69 @@
+// Copyright (c) 2023 Samsung Electronics Co., Ltd. All Rights Reserved.
```
Comment (gapisback) [inline third_party/islet/include/islet.h:1]: Oh, I see this copyright here. Do you [not] need to mention the license under which this is copyrighted?
Similar copyright notice may be needed in `.cc`, `.c` file(s).

File: third_party/islet/include/islet.h [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+
+constexpr static const int ISLET_E_FAILURE = -1;
+
+constexpr static const int ISLET_E_INPUT = -2;
+
+constexpr static const int ISLET_E_WRONG_REPORT = -3;
+
+constexpr static const int ISLET_E_WRONG_CLAIMS = -4;
+
+constexpr static const int ISLET_E_NOT_SUPPORTED = -5;
```
Comment (gapisback) [inline third_party/islet/include/islet.h:21]: Is it possible to rewrite these constants as enum-values?

Say, you define a typedef `islet_rv_t`, and define all these as this enumerated type? Not sure if this is feasible in this C++ code.

Enum-types are helpful as compiler will help checking for returned value. Also, you can see the enum-type-name displayed while debugging.

File: third_party/islet/include/islet.h [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+
+extern "C" {
+
+/// Get an attestation report(token).
+///
+/// # Note
+/// This API returns hard coded report to simuate temporarily.
+/// `User data` could be used as nonce to prevent reply attack.
+int islet_attest(const unsigned char *user_data,
```
Comment (gapisback) [inline third_party/islet/include/islet.h:30]: If you do choose the suggestion of enum-types, I think all these functions' definition should change to return `islet_rv_t`.

File: third_party/islet/include/islet.h [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+                unsigned char *value_out,
+                int *value_out_len);
+
+/// Print all claims including Realm Token and Platform Token.
+void islet_print_claims(const unsigned char *claims, int claims_len);
+
+/// Seals the plaintext given into the binary slice
+///
+/// # Note
+/// This API seals with hard coded key to simuate temporarily.
```
Comment (gapisback) [inline third_party/islet/include/islet.h:54]: Typo: 'simulate'.

I think a rephrased message would be better:

```
This API currently seals with a hard-coded key, to simulate seal operation. 
In future, this will be finalized to support _<fill-in-the-blank>_ keys.
```

Replace the _<fill-in-the-blank>_  with some term better describing what you intend to do in the future.

Hope this rewrite makes sense to you ...

File: third_party/islet/include/islet.h [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+int islet_seal(const unsigned char *plaintext,
+               int plaintext_len,
+               unsigned char *sealed_out,
+               int *sealed_out_len);
+
+/// Unseals the sealed bianry given into the plaintext
```
Comment (gapisback) [inline third_party/islet/include/islet.h:60]: Typo: 'binary'. 

Maybe rephrase as: `Unseals into plaintext the sealed binary provided.`

File: third_party/islet/include/islet.h [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+///
+/// # Note
+/// This API unseals with hard coded key to simuate temporarily.
```
Comment (gapisback) [inline third_party/islet/include/islet.h:63]: Typo: 'simulate'.

You may want to rephrase this comment, also, based on the suggestion given on L54 for the seals API comment.

File: third_party/islet/setup.sh [hunk 10]
```diff
@@ -0,0 +1,30 @@
+#!/bin/bash
+
```
Comment (gapisback) [inline third_party/islet/setup.sh:2]: Please add a small banner prolog comment documenting what this shell script is for / about.

You may want to follow the convention introduced (by me) recently in `sample_apps/run_example.sh`.

File: third_party/islet/setup.sh [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+set -e
```
Comment (gapisback) [inline third_party/islet/setup.sh:3]: This is good, but over time, ... I have been taught (& discovered) that a more stricter checking for errors during shell script execution is better.

See if you can expand this to `set -Eeuo pipefail`, so that any errors in any command will cause this script to abort immediately.

This business of `$rc` error checking in bash scripts is bit difficult to work with. I experimented with some of this most recently in above-mentioned shell script. You may want to review what I did, to see if that model will work for your script. And adapt that accordingly.

Comment (gapisback) [reply third_party/islet/setup.sh:3]: Also it is a good idea to always run `shellcheck` on this script, and fix all warnings / errors reported. Have you done that?

File: third_party/islet/setup.sh [hunk 12]
```diff
@@ -0,0 +1,30 @@
...
+
+ROOT=$(git rev-parse --show-toplevel)
```
Comment (gapisback) [inline third_party/islet/setup.sh:5]: Is this better named as `CC_ROOT`, for clarity that this is the top-level git-sub-dir for CC library?

(Or, if this is some other repo, consider adding a prefix to indicate what this `root` dir is for.)

File: third_party/islet/setup.sh [hunk 13]
```diff
@@ -0,0 +1,30 @@
...
+
+# Install rust
+$ISLET/scripts/deps/rust.sh
+
+# Build islet sdk (simulated version for x86_64)
+cd $ISLET_SDK && cargo build
+
+mkdir -p $ISLET_INC $ISLET_LIB
+cp $TARGET_HDR $ISLET_INC
+cp $TARGET_LIB $ISLET_LIB
```
Comment (gapisback) [inline third_party/islet/setup.sh:30]: Please try to always use `cp -p` so that the file's timestamp is preserved. (Helps script debugging.)

**Review by gapisback — COMMENTED** [review]
Hi, @bitboom -- I am a developer on the CC library with VMware.

I wanted to discover and learn what your team has done with this Islet integration.

So, I took the opportunity to review this code based on our discussion in the team meeting last night, with @jlmucb .

I left my review as 'open', expecting that you will study the comments and adjust accordingly. Once John gets to see this, feel free to discuss with him how far to go toward implementing the suggestions given. (Most of my changes are stylistic, and clean-up type, so don't expect that they will take too much effort to update the code.)

Thanks.
--AdityA>

**Comment by gapisback** [comment]
Attn: @bitboom -- I see you have 3 commits wanting to merge to `/main`. Once the code reviews are done, please do remember to squash all commits into one, and add a single commit-message.

In the UI, you should be able to pick "Squash and rebase".

Or, you can do squash offline in your dev-branch, and update the commit-message to describe briefly the changes done.
Then force-push this back to your dev-branch and pick "Rebase and merge" (or some such button) in this UI to push this to `/main`.

File: sample_apps/islet_attest_seal/attest_seal_app.cc [hunk 14]
```diff
@@ -0,0 +1,79 @@
...
+
+static const char CLAIM_TITLE_USER_DATA[] = "User data";
+static const char CLAIM_TITLE_PLATFORM_PROFILE[] = "Profile";
+
+bool attestation_test() {
+    byte report[2048], claims[1024], value[1024];
+    memset(report, 0, sizeof(report));
+    memset(claims, 0, sizeof(report));
+    memset(value, 0, sizeof(value));
+    int report_len = 0, claims_len = 0, value_len = 0;
```
[the remark this answers is not in the document]
Comment (jlmucb) [reply sample_apps/islet_attest_seal/attest_seal_app.cc:17]: yes.   They c++ style is generally the "Google C++" style which Google has published

File: sample_apps/islet_attest_seal/attest_seal_app.cc [hunk 15]
```diff
@@ -0,0 +1,79 @@
+#include <islet.h>
```
[the remark this answers is not in the document]
Comment (jlmucb) [reply sample_apps/islet_attest_seal/attest_seal_app.cc:1]: The copyright notice for Samsung contributions should be (c) Samsung and say it is licensed under the Apache license (or if another license, it shoudl say and include the license in thedirectory).

File: sample_apps/islet_attest_seal/attest_seal_app.cc (hunk 14 above)
[the remark this answers is not in the document]
Comment (jlmucb) [reply sample_apps/islet_attest_seal/attest_seal_app.cc:17]: Incidentially, in the Google C++ style, the "tab" is 2 spaces, not 4.

File: third_party/islet/setup.sh (hunk 13 above)
Comment (bitboom) [reply third_party/islet/setup.sh]: I attached `-p` option.

File: third_party/islet/setup.sh (hunk 12 above)
Comment (bitboom) [reply third_party/islet/setup.sh]: I changed it to `CC_ROOT`.

File: third_party/islet/include/islet.h (hunk 9 above)
Comment (bitboom) [reply third_party/islet/include/islet.h]: I fixed it. Thanks!

File: sample_apps/islet_attest_seal/attest_seal_app.cc [hunk 16]
```diff
@@ -0,0 +1,79 @@
...
+#include <string>
+#include <cstring>
+
+using byte = unsigned char;
+
+static const char CLAIM_TITLE_USER_DATA[] = "User data";
+static const char CLAIM_TITLE_PLATFORM_PROFILE[] = "Profile";
+
+bool attestation_test() {
+    byte report[2048], claims[1024], value[1024];
```
[the remark this answers is not in the document]
Comment (bitboom) [reply sample_apps/islet_attest_seal/attest_seal_app.cc]: I defined `BUFFER_SIZE` :)

File: sample_apps/islet_attest_seal/README.md [hunk 17]
```diff
@@ -0,0 +1,21 @@
+# A sample app towards Arm CCA
+This sample app is towards Arm CCA with [ISLET SDK](https://github.com/Samsung/islet/tree/main/sdk)
+Currently this is running on x86_64 to intergrate easily. 
+The real app running on Arm CCA simulator (arm64) is comming soon.
+
+## Getting started
+### Setup islet sdk
+ISLET SDK is written in rust but it also support C/C++.
```
[the remark this answers is not in the document]
Comment (bitboom) [reply sample_apps/islet_attest_seal/README.md]: Thanks for suggestion! I fixed it!

File: sample_apps/islet_attest_seal/README.md [hunk 18]
```diff
@@ -0,0 +1,21 @@
+# A sample app towards Arm CCA
+This sample app is towards Arm CCA with [ISLET SDK](https://github.com/Samsung/islet/tree/main/sdk)
+Currently this is running on x86_64 to intergrate easily. 
+The real app running on Arm CCA simulator (arm64) is comming soon.
```
[the remark this answers is not in the document]
Comment (bitboom) [reply sample_apps/islet_attest_seal/README.md]: I fixed it.

File: sample_apps/islet_attest_seal/README.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (bitboom) [reply sample_apps/islet_attest_seal/README.md]: Thanks!

File: sample_apps/islet_attest_seal/attest_seal_app.cc (hunk 14 above)
[the remark this answers is not in the document]
Comment (bitboom) [reply sample_apps/islet_attest_seal/attest_seal_app.cc]: I formatted it by `astyle --style=google --indent=spaces=2`. Thanks for letting me know Coding Style what certifier followed :+1:

File: sample_apps/islet_attest_seal/attest_seal_app.cc [hunk 19]
```diff
@@ -0,0 +1,79 @@
...
+    if (islet_parse(CLAIM_TITLE_PLATFORM_PROFILE, claims, claims_len, value, &value_len))
+        return false;
+
+    printf("Claim[Platform  profile]: %s\n", (char*) value);
+
+    return true;
+}
+
+bool sealing_test() {
+    byte sealed[2048], unsealed[2048];
```
[the remark this answers is not in the document]
Comment (bitboom) [reply sample_apps/islet_attest_seal/attest_seal_app.cc]: I patched it. Thanks!

File: sample_apps/islet_attest_seal/attest_seal_app.cc [hunk 20]
```diff
@@ -0,0 +1,79 @@
...
+
+    return true;
+}
+
+int main() {
+    printf("[!] Starting attestation test...\n");
+    if (attestation_test()) 
+        printf("[!] Success to attestation test...\n");
+    else
+        printf("[!] Failed to attestation test...\n");
```
[the remark this answers is not in the document]
Comment (bitboom) [reply sample_apps/islet_attest_seal/attest_seal_app.cc]: It looks more concise. Thanks!

File: sample_apps/islet_attest_seal/attest_seal_app.cc (hunk 2 above)
Comment (bitboom) [reply sample_apps/islet_attest_seal/attest_seal_app.cc]: The reason is that those are orthogonal test. But i think it is better to exit when `attestation_test()` is failed.

File: third_party/islet/include/islet.h (hunk 5 above)
Comment (bitboom) [reply third_party/islet/include/islet.h]: Thanks for suggestion! I added `islet_status_t` enum :)

File: third_party/islet/include/islet.h (hunk 6 above)
Comment (bitboom) [reply third_party/islet/include/islet.h]: I fixed it :)

File: third_party/islet/include/islet.h (hunk 7 above)
Comment (bitboom) [reply third_party/islet/include/islet.h]: Thanks for suggestion! I reflected it.

File: third_party/islet/include/islet.h (hunk 8 above)
Comment (bitboom) [reply third_party/islet/include/islet.h]: I reflected it! Thanks.

File: third_party/islet/setup.sh (hunk 11 above)
Comment (bitboom) [reply third_party/islet/setup.sh]: I didn't use `shellcheck`. (I actually have used `shfmt` on my projects).
Thanks for valuable comments! I reflected both comments.

**Comment by bitboom** [comment]
> Attn: @bitboom -- I see you have 3 commits wanting to merge to `/main`. Once the code reviews are done, please do remember to squash all commits into one, and add a single commit-message.
> 
> In the UI, you should be able to pick "Squash and rebase".

@gapisback 
I will squash commits into one by using UI as you mentioned.
And i reflected your comments. Really thank you for kind comments!

**Comment by bitboom** [comment]
@jlmucb 
I received the email and thank you so much for the warm and friendly welcome 😄.
I replied to your comments below. 

Please let me know if there are other comments.

---

> There are a bunch of comments we left in the github repository.  The main comment is that this will turn into interface code and interface code test, not a "sample app,"  so it should go in a new directory $CERTIFIER/src/cca, for example.  Later, there will be a directory in sample_apps called simple_app_under_cca and maybe other sample apps you contribute.

I moved interface code test moved to `src/cca/islet_test`.

> A minor comment:  The C++ style we use is the Google style (we use Google Go style too).  It's easy to find on the web.  2 spaces for tab, usually single variable declaration per line although we're not dictatorial.  

I reflected style with `astyle --style=google --indent=spaces=2 *.cc` . And I also fixed it to `single variable declaration`.

> The files you contribute in their entirety should bear a Samsung copyright notice and say they are licensed under the Apache license in the main directory.  If you want to license under BSD, you can do that but we should include a copy of that license in the main directory, fo reference in that case.

I added a Samsung Copyright notice under the Apache license  to `*.h,  *.cc`.

**Comment by bitboom** [comment]
@jlmucb I have finally done CLA process. :smile: 
Could you please check this PR can be merged?

---

FYI, I recently done to get `real attestation report` from `CCA simulator` at https://github.com/Samsung/islet/pull/76.
So, i will make next PR to certifier which is run `islet test` on CCA simulator.

File: src/cca/islet_test/attest_seal_test.cc [hunk 21]
```diff
@@ -0,0 +1,102 @@
...
+ *  limitations under the License
+ */
+
+#include <islet.h>
+
+#include <iostream>
+#include <string>
+#include <cstring>
+
+#define BUFFER_SIZE 2048
```
Comment (gapisback) [inline src/cca/islet_test/attest_seal_test.cc:23]: Please, can you document where we arrived at this limit from? The reason for these kinds of sizes will be lost over time, so even if you can add a simple comment, something like:

// Some reasonable size to allocate <... fill-in-the-blank> ... on-stack buffers.

If you can give the reason why 2K is big enough, that will be helpful. (For future readers.)

File: src/cca/islet_test/attest_seal_test.cc [hunk 22]
```diff
@@ -0,0 +1,102 @@
...
+bool attestation_test() {
+  byte report[BUFFER_SIZE];
+  byte claims[BUFFER_SIZE];
+  byte value[BUFFER_SIZE];
+  int report_len = 0;
+  int claims_len = 0;
+  int value_len = 0;
+
+  memset(report, 0, sizeof(report));
+  memset(claims, 0, sizeof(report));
```
Comment (gapisback) [inline src/cca/islet_test/attest_seal_test.cc:39]: Although this will work, technically this should be : `sizeof(claims)`. Please fix.

File: third_party/islet/include/islet.h [hunk 23]
```diff
@@ -0,0 +1,83 @@
...
+/// Get an attestation report(token).
+///
+/// # Note
+/// This API currently returns hard-coded report to simulate attest operation.
+/// In future, this will be finalized to support reports signed by RMM.
+/// `User data` could be used as nonce to prevent reply attack.
+islet_status_t islet_attest(const unsigned char *user_data,
+                            int user_data_len,
+                            unsigned char *report_out,
+                            int *report_out_len);
```
Comment (gapisback) [inline third_party/islet/include/islet.h:45]: Thank you for the nice indentation of fn-arguments. It makes the `.h` very much more readable.

File: third_party/islet/setup.sh [hunk 24]
```diff
@@ -0,0 +1,34 @@
+#!/bin/bash
+# ##############################################################################
+# setup.sh - Setup script to build ISLET SDK
+# ##############################################################################
+
+set -Eeuo pipefail
```
Comment (gapisback) [inline third_party/islet/setup.sh:6]: (FYI: I'm assuming you've run `shellcheck` on this script. Right now, there are only a few Bash scripts. I am thinking it will be good to invoke `shellcheck` as part of CI, so that every new script change will adhere to common formatting. 

That's not in-place, just yet ... but will be good if have run this thru `shellcheck` while we are adding a new script.
)

**Review by gapisback — APPROVED** [review]
Took a quick re-look at your updated files. I've put down few very minor comments, which, if you can address, it will be good.

Thanks for improving the code based on my previous set of minor comments. Things are beginning to look more polished with this rev.

File: src/cca/islet_test/attest_seal_test.cc (hunk 21 above)
Comment (bitboom) [reply src/cca/islet_test/attest_seal_test.cc:25]: I added it. Thank you for the suggestion!

File: third_party/islet/setup.sh [hunk 25]
```diff
@@ -0,0 +1,34 @@
+#!/bin/bash
+# ##############################################################################
+# setup.sh - Setup script to build ISLET SDK
+# ##############################################################################
```
[the remark this answers is not in the document]
Comment (bitboom) [reply third_party/islet/setup.sh:19]: Thank you for pointing out. I added the copyright.

File: third_party/islet/setup.sh (hunk 24 above)
Comment (bitboom) [reply third_party/islet/setup.sh:21]: Yes. It's very good to add `shellcheck` to CI. 
And  c++ conding style check on CI is also good too =)

File: src/cca/islet_test/attest_seal_test.cc (hunk 22 above)
Comment (bitboom) [reply src/cca/islet_test/attest_seal_test.cc]: You're right. I fixed it.

**Comment by bitboom** [comment]
> Can you remove the build.yml changes (but save the file) from the pull request? Were tidying up the yml tests. If you do that, you can merge (or I'll merge for you). Everything else is fine.
> 
> Thanks

I removed it =)

**Review by jlmucb — APPROVED** [review]

**Review by jlmucb — APPROVED** [review]
